pkg/proxy/middleware.go

```diff
@@ -0,0 +1,123 @@
+package proxy
+
+import (
+	"io"
+	"math/rand"
+	"net/http"
+	"time"
+)
+
+var _ http.Handler = &RetryMiddleware{}
+
+type RetryMiddleware struct {
+	nextHandler http.Handler
+	MaxRetries  int
+	rSource     *rand.Rand
+}
+
+func NewRetryMiddleware(maxRetries int, other http.Handler) *RetryMiddleware {
+	if maxRetries < 1 {
+		panic("invalid retries")
+	}
+	return &RetryMiddleware{
+		nextHandler: other,
+		MaxRetries:  maxRetries,
+		rSource:     rand.New(rand.NewSource(time.Now().UnixNano())),
+	}
+}
+
+func (r RetryMiddleware) ServeHTTP(writer http.ResponseWriter, request *http.Request) {
+	var capturedResp *responseWriterDelegator
+	for i := 0; ; i++ {
+		capturedResp = &responseWriterDelegator{
+			ResponseWriter: writer,
+			headerBuf:      make(http.Header),
+			discardErrResp: i < r.MaxRetries &&
+				request.Context().Err() == nil, // abort early on timeout, context cancel
+		}
+		// call next handler in chain
+		r.nextHandler.ServeHTTP(capturedResp, request.Clone(request.Context()))
+
+		if !capturedResp.discardErrResp || // max retries reached
+			!isRetryableStatusCode(capturedResp.statusCode) {
+			break
+		}
+		totalRetries.Inc()
+		// Exponential backoff
+		jitter := time.Duration(r.rSource.Intn(50))
+		time.Sleep(time.Millisecond*time.Duration(1<<uint(i)) + jitter)
+	}
+}
+
+func isRetryableStatusCode(status int) bool {
+	return status == http.StatusBadGateway ||
```

Comment:
the values are still hardcoded here. I will make them configurable. 
Not sure about [`408`](https://http.dev/408). When the client is too slow, would a retry help?

Reply:
I am good with leaving off 408